kale/pipeline/action_domain_adapter.py

```diff
@@ -0,0 +1,673 @@
+
+        elif self.image_modality == "joint":
+            set_requires_grad(self.rgb_feat, requires_grad=False)
+            set_requires_grad(self.flow_feat, requires_grad=False)
+            (x_s_rgb, y_s), (x_s_flow, _), (x_tu_rgb, _), (x_tu_flow, _) = batch
+            with torch.no_grad():
+                h_s_rgb = self.rgb_feat(x_s_rgb).data.view(x_s_rgb.shape[0], -1)
+                h_t_rgb = self.rgb_feat(x_tu_rgb).data.view(x_tu_rgb.shape[0], -1)
+                h_s_flow = self.flow_feat(x_s_flow).data.view(x_s_flow.shape[0], -1)
+                h_t_flow = self.flow_feat(x_tu_flow).data.view(x_tu_flow.shape[0], -1)
+                h_s = torch.cat((h_s_rgb, h_s_flow), dim=1)
+                h_t = torch.cat((h_t_rgb, h_t_flow), dim=1)
+
+            for _ in range(self._k_critic):
+                gp = losses.gradient_penalty(self.domain_classifier, h_s, h_t)
+
+                critic_s = self.domain_classifier(h_s)
+                critic_t = self.domain_classifier(h_t)
+                wasserstein_distance = critic_s.mean() - (1 + self._beta_ratio) * critic_t.mean()
+
+                critic_cost = -wasserstein_distance + self._gamma * gp
+
+                self.critic_opt.zero_grad()
+                critic_cost.backward()
+                self.critic_opt.step()
+                if self.critic_sched:
+                    self.critic_sched.step()
+
+                # Uncomment for later work. Process rgb and flow dividedly.
+                # for _ in range(self._k_critic):
+                #     gp_rgb = losses.gradient_penalty(self.domain_classifier, h_s_rgb, h_t_rgb)
+                #     gp_flow = losses.gradient_penalty(self.domain_classifier, h_s_flow, h_t_flow)
+                #
+                #     critic_s_rgb = self.domain_classifier(h_s_rgb)
+                #     critic_s_flow = self.domain_classifier(h_s_flow)
+                #     critic_t_rgb = self.domain_classifier(h_t_rgb)
+                #     critic_t_flow = self.domain_classifier(h_t_flow)
+                #     wasserstein_distance_rgb = (
+                #             critic_s_rgb.mean() - (1 + self._beta_ratio) * critic_t_rgb.mean()
+                #     )
+                #     wasserstein_distance_flow = (
+                #             critic_s_flow.mean() - (1 + self._beta_ratio) * critic_t_flow.mean()
+                #     )
+                #
+                #     critic_cost = (-wasserstein_distance_rgb + -wasserstein_distance_flow +
+                #                    self._gamma * gp_rgb + self._gamma * gp_flow) * 0.5
+
+                self.critic_opt.zero_grad()
+                critic_cost.backward()
+                self.critic_opt.step()
+                if self.critic_sched:
+                    self.critic_sched.step()
+
+            set_requires_grad(self.rgb_feat, requires_grad=True)
+            set_requires_grad(self.flow_feat, requires_grad=True)
+            set_requires_grad(self.domain_classifier, requires_grad=False)
```

Comment:
I did not find major problems above. Tests will be helpful. Also it may be possible to reduce redundancy.